integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+import static javax.transaction.xa.XAResource.TMRESUME;
+import static javax.transaction.xa.XAResource.TMSTARTRSCAN;
+import static javax.transaction.xa.XAResource.TMSUCCESS;
+import static javax.transaction.xa.XAResource.TMSUSPEND;
+import static javax.transaction.xa.XAResource.XA_OK;
+import static javax.transaction.xa.XAResource.XA_RDONLY;
+
+/**
+ * An {@link XAResource} that adapts an ordinary arbitrary {@link Connection} as much as possible to the {@link
+ * XAResource} contract.
+ *
+ * <p><strong>Note:</strong> instances of this class are lossless in the presence of one-phase commit operations and
+ * potentially lossy in the presence of two-phase commit operations.</p>
+ *
+ * <p>Instances of this class are safe for concurrent use by multiple threads.</p>
+ */
+final class LocalXAResource implements XAResource {
+
+
+    /*
+     * Static fields.
+     */
+
+
+    private static final Logger LOGGER = Logger.getLogger(LocalXAResource.class.getName());
+
+    private static final Xid[] EMPTY_XID_ARRAY = new Xid[0];
+
+    // package-protected for testing only.
+    static final ConcurrentMap<Xid, Association> ASSOCIATIONS = new ConcurrentHashMap<>();
+
+
+    /*
+     * Instance fields.
+     */
+
+
+    private final Function<? super Xid, ? extends Connection> connectionFunction;
+
+    private final ExceptionConverter exceptionConverter;
+
+
+    /*
+     * Constructors.
+     */
+
+
+    /**
+     * Creates a new {@link LocalXAResource}.
+     *
+     * @param connectionFunction a {@link Function} that accepts a {@link Xid} (supplied by the {@link #start(Xid, int)}
+     * method) and returns a {@link Connection} to associate with the global transaction; must not be {@code null}; must
+     * never return {@code null}; must be safe for concurrent use by multiple threads; will never be invoked with a
+     * {@code null} {@link Xid}
+     *
+     * @param exceptionConverter a {@link ExceptionConverter} that accepts a {@link XARoutine} and a {@link
+     * SQLException} and converts the {@link SQLException} to an <em>appropriate</em> {@link XAException} following the
+     * rules defined by the <a href="https://pubs.opengroup.org/onlinepubs/009680699/toc.pdf">XA Specification</a> as
+     * interpreted by the specification of the {@code javax.transaction.xa} package and its classes; may be {@code null}
+     * in which case a default implementation will be used instead
+     *
+     * @see #start(Xid, int)
+     */
+    LocalXAResource(Function<? super Xid, ? extends Connection> connectionFunction, ExceptionConverter exceptionConverter) {
+        super();
+        this.connectionFunction = Objects.requireNonNull(connectionFunction, "connectionFunction");
+        this.exceptionConverter = exceptionConverter == null ? LocalXAResource::convert0 : exceptionConverter;
+    }
+
+
+    /*
+     * Instance methods.
+     */
+
+
+    @Override // XAResource
+    public void start(Xid xid, int flags) throws XAException {
```

Comment:
The XA protocol—the permitted sequence of invocations on an `XAResource`—begins with `start`.  `start`'s job is to mark the beginning of the association of operations on the current thread against the "resource manager" (the database connection in this case) with the global transaction identified by the supplied `Xid`.  (The ending of this thread association is accomplished by the `end` method.)